pkg/bff/collaborators.go

```diff
@@ -65,74 +64,60 @@ func (s *Server) AddCollaborator(c *gin.Context) {
-		c.JSON(http.StatusForbidden, api.ErrorResponse("maximum number of collaborators reached"))
-		return
-	}
-
-	// Don't overwrite an existing collaborator
-	id := collaborator.Key()
-	if _, ok := org.Collaborators[id]; ok {
-		log.Warn().Str("collabID", id).Msg("collaborator already exists")
-		c.JSON(http.StatusConflict, api.ErrorResponse("collaborator already exists"))
-		return
-	}
-
-	// Make sure the record has a created timestamp
-	if collaborator.CreatedAt == "" {
-		collaborator.CreatedAt = time.Now().UTC().Format(time.RFC3339Nano)
-	}
+	// Collaborator invites are valid for seven days
+	collaborator.ExpiresAt = time.Now().AddDate(0, 0, 7).Format(time.RFC3339Nano)
 
 	// Add the collaborator to the organization
-	if org.Collaborators == nil {
-		org.Collaborators = make(map[string]*models.Collaborator)
+	if err = org.AddCollaborator(collaborator); err != nil {
+		switch {
+		case errors.Is(err, models.ErrInvalidCollaborator):
+			log.Warn().Err(err).Msg("invalid collaborator")
+			c.JSON(http.StatusBadRequest, api.ErrorResponse(err))
+		case errors.Is(err, models.ErrCollaboratorExists):
+			log.Warn().Err(err).Str("email", collaborator.Email).Msg("collaborator already exists")
+			c.JSON(http.StatusConflict, api.ErrorResponse(err))
+		case errors.Is(err, models.ErrMaxCollaborators):
+			log.Warn().Err(err).Int("maximum", models.MaxCollaborators).Msg("maximum number of collaborators reached")
+			c.JSON(http.StatusForbidden, api.ErrorResponse(err))
+		default:
+			log.Error().Err(err).Msg("could not add collaborator")
+			c.JSON(http.StatusInternalServerError, api.ErrorResponse(err))
+		}
+		return
 	}
-	org.Collaborators[id] = collaborator
 
-	// If the user doesn't exist in Auth0 then we need to create them
+	// Handle both users who have already registered and completely new users
 	var user *management.User
+	var baseURL string
 	if user, err = s.FindUserByEmail(collaborator.Email); err != nil {
 		if !errors.Is(err, ErrUserEmailNotFound) {
 			log.Error().Err(err).Str("email", collaborator.Email).Msg("error finding user by email in Auth0")
 			c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not add collaborator"))
 			return
 		}
 
-		// Create an unverified user in Auth0
-		// Note: If the user has already registered with another IDP then they will need to link accounts
-		var verifyEmail bool
-		password := secrets.CreateToken(32)
+		// If the user doesn't exist in Auth0 then direct them to the signup page
+		baseURL = s.conf.RegisterURL
 		user = &management.User{
-			Email:       &collaborator.Email,
-			Password:    &password,
-			Connection:  &s.conf.Auth0.ConnectionName,
-			VerifyEmail: &verifyEmail,
-		}
-		if err = s.auth0.User.Create(user); err != nil {
-			log.Error().Err(err).Str("email", collaborator.Email).Msg("error creating user in Auth0")
-			c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not add collaborator"))
-			return
+			Email: &collaborator.Email,
```

Comment:
Previously we were creating the Auth0 user on the user's behalf (normally it would happen when they sign up on the frontend) if they didn't exist yet. However, this just complicated things because it involved maintaining state in both the BFF and Auth0 and it was difficult to tell which users had verified their email. To simplify this, the new code just creates the collaborator record in Auth0 and all users still have to go through the standard registration process to create their Auth0 account.

Reply:
Do we need an additional test to cover the case where an existing Auth0 user without an org now needs to log in and become associated with the org?

Reply:
I think this is covered in the "user invite" test - we have a case where the user is logging in with an `orgid` in the request but has no current `orgid` in the Auth0 metadata - in which case we have to check the database to see if they were invited. https://github.com/trisacrypto/directory/blob/cfd08c2be9d8d2f0002e17dc724e65bd6d4f4c35/pkg/bff/users_test.go#L196